Simplify the steps to make changes to config.json

krishnasrinivas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Reduce the code that needs to be touched when there is config version bump.
Have well defined steps now:
```
+// To introduce a new config version:
+// 1. Change the configCurrentVersion
+// 2. Change the necessary serverConfigCurrent fields
+// 3. Add new migration function (ex. migrateV19ToV20())
+// 4. Add migration calling code to migrateConfig() in config-migrate.go
+// 5. Make changes in config-current_test.go for any test change
```

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
go test

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

krishnasrinivas committed:
Simplify the steps to make changes to config.json

krisis commented on cmd/admin-handlers.go:
```diff
@@ -985,7 +985,7 @@ func (adminAPI adminAPIHandlers) SetConfigHandler(w http.ResponseWriter, r *http
 	}
 
-	var config serverConfigV20
+	var config serverConfigCurrent
```
With this approach, how will be able to import config file with version 20 when we move to version 21? IIUC, when we release minio with config version 21, we won't have a type which is capable of importing config file version 20 during migration.

krishnasrinivas replied:
Added additional step:
```
Copy serverConfigCurrent struct to serverConfigVN (in config-old.go)
```

Check the new steps:
```
// Steps to move from version N to version N+1
// 1. Copy serverConfigCurrent struct to serverConfigVN (in config-old.go)
// 2. Change the necessary serverConfigCurrent fields to accomodate new changes.
// 3. Change the configCurrentVersion to "N+1"
// 4. Add new migration function (ex. migrateVNToVN+1())
// 5. Add migrateVNToVN+1() to migrateConfig() in config-migrate.go
// 6. Make changes in config-current_test.go for any test change
```

krisis replied:
To verify this steps in practice, we could move serverConfig19 to config-old.go and follow the remaining steps explained in the above comment. This would leave a working example besides the above comments for future developers.

harshavardhana replied:
@krishnasrinivas can you incorporate @krisis recommendation? - We can move ahead and merge this PR with that change in place.

poornas reviewed (approved):
Tested locally - looks good

kannappanr commented on cmd/config-current.go:
```diff
@@ -27,44 +27,35 @@ import (
+
 // Config version
-const v20 = "20"
+const configCurrentVersion = "20"
```
Do you think the variable name can be changed from ``` configCurrentVersion ``` to ``` currentConfigversion ``` ?

krishnasrinivas replied:
will change it to `serverConfigVersion`

kannappanr commented on cmd/config-current.go:
```diff
@@ -27,44 +27,35 @@ import (
-const v20 = "20"
+const configCurrentVersion = "20"
+
+type serverConfigCurrent = serverConfigV20
```
variable name ``` serverConfigCurrent ``` can be changed to either ``` serverConfig ``` or ``` currentServerConfig ```

krishnasrinivas replied:
yes, good suggestion, will change it to serverConfig

krishnasrinivas committed:
Rename:
configCurrentVersion->serverConfigVersion
serverConfigCurrent->serverConfig
serverConfig->globalServerConfig
serverConfigMu->globalServerConfigMu

kannappanr reviewed (approved):
LGTM
